src/components/MessageInput/hooks/useCooldownTimer.tsx

```diff
@@ -53,29 +41,32 @@ export type CooldownTimerState = {
 export const useCooldownTimer = <
   StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
 >(): CooldownTimerState => {
-  const { client } = useChatContext<StreamChatGenerics>('useCooldownTimer');
+  const { latestMessageDatesByChannels } = useChatContext<StreamChatGenerics>('useCooldownTimer');
   const { channel } = useChannelStateContext<StreamChatGenerics>('useCooldownTimer');
 
-  const { cooldown: cooldownInterval } = (channel.data ||
+  const { cooldown: cooldownInterval, own_capabilities } = (channel.data ||
     {}) as ChannelResponse<StreamChatGenerics>;
 
   const [cooldownRemaining, setCooldownRemaining] = useState<number>();
 
-  const globalRole = client.user?.role || '';
-  const channelRole = channel.state.members[client.userID || '']?.role || '';
-
-  const skipCooldown = checkUserRoles(globalRole, channelRole);
+  const skipCooldown = !own_capabilities?.includes('slow-mode');
 
   useEffect(() => {
-    const handleEvent = (event: Event<StreamChatGenerics>) => {
-      if (!skipCooldown && event.user?.id === client.userID) {
-        setCooldownRemaining(cooldownInterval);
-      }
-    };
-
-    if (cooldownInterval) channel.on('message.new', handleEvent);
-    return () => channel.off('message.new', handleEvent);
-  }, [channel.id, cooldownInterval]);
-
-  return { cooldownInterval: cooldownInterval || 0, cooldownRemaining, setCooldownRemaining };
+    const latestMessageDate = latestMessageDatesByChannels[channel.cid];
+    if (!cooldownInterval || !latestMessageDate) {
+      return;
+    }
+    const remainingCooldown = Math.round(
+      cooldownInterval - (new Date().getTime() - latestMessageDate.getTime()) / 1000,
+    );
+    if (remainingCooldown > 0 && !skipCooldown) {
+      setCooldownRemaining(remainingCooldown);
+    }
+  }, [channel.id, cooldownInterval, latestMessageDatesByChannels[channel.cid]]);
+
+  return {
+    cooldownInterval: cooldownRemaining || cooldownInterval || 0,
```

Comment:
I was looking into the use of cooldownInterval and it seems that you always want cooldownRemaining. But maybe I am missing something?

Reply:
Well, this bit is a bit of hack, previously the `cooldownInterval` was the fixed value of the cooldown period, and `cooldownRemaining` was the value that showed how much time is remaining in the current timer.

The problem was that `cooldownInterval` was used as a prop to the CooldownTimer component. But this was not the right behavior because if you switch channels while during the cooldown, the cooldown should not start from `cooldownInterval` but rather from `cooldownRemaining`.

But since `cooldownInterval` originally means the fixed value of the timeout, it seemed wrong to always return cooldownRemaining since that could be 0.

The proper solution would've been to use `cooldownRemaining` as the prop for the CooldownTimer component, but I was too chicken to update everywhere, not being sure if that would cause issues elsewhere.

I'm not sure if this was clear or not.

Reply:
Hm, after seeing how much explanation this code snippet needed, I think it's best if I choose the right solution, not the hacky one and update the props of the `CooldownTimer` components.

Reply:
@MartinCupela done